services/httpd/limiter.go

```diff
@@ -0,0 +1,54 @@
+package httpd
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+
+	"github.com/influxdata/influxdb/influxql"
+)
+
+// ErrConnectionLimitExceeded indicates the connection limit has been reached.
+var ErrConnectionLimitExceeded = errors.New("connection limit exceeded")
+
+// connectionLimiter limits the maximum number of connections.
+type connectionLimiter struct {
+	active uint64
+	limit  uint64
+	mu     sync.Mutex
+}
+
+// wrap wraps an existing handler to increment and decrement the active
+// connections and then invoke the handler. If it cannot increment the number
+// of active connections because the limit has been exceeded, it sends an HTTP
+// response immediately to close the connection.
+func (l *connectionLimiter) wrap(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if err := l.increment(); err != nil {
+			w.Header().Set("Connection", "close")
+			resultError(w, influxql.Result{Err: err}, http.StatusBadRequest)
+			return
+		}
+		defer l.decrement()
+		h.ServeHTTP(w, r)
+	})
+}
+
+// increment increments the number of active connections. It returns an error
+// if the connection limit is exceeded.
+func (l *connectionLimiter) increment() error {
+	l.mu.Lock()
```

Comment:
Every `mu.Lock` in the write path has had to be reworked at some point due to contention.  We definitely need to benchmark/profile this.